src/savegame_json.cpp

```diff
@@ -1115,19 +1115,45 @@ void monster::load(JsonObject &data)
 
     data.read("hp", hp);
 
+    // sp_timeout indicates an old save, prior to the special_attacks refactor
     if (data.has_array("sp_timeout")) {
         JsonArray parray = data.get_array("sp_timeout");
         if ( !parray.empty() ) {
+            int index = 0;
             int ptimeout = 0;
-            while ( parray.has_more() ) {
+            while ( parray.has_more() && index < (int)(type->special_attacks_names.size()) ) {
                 if ( parray.read_next(ptimeout) ) {
-                    sp_timeout.push_back(ptimeout);
+                    // assume timeouts saved in same order as current monsters.json listing
+                    std::string aname = type->special_attacks_names[index++];
+                    if ( ptimeout >= 0 ) {
```

Comment:
If you don't like the constructor usage of `mon_special_attack`, what about this:
```C++
class mon_special_attack {
public:
	int timeout = 0; // default value
	bool enabled = true;
};

void load_from_json() {
	auto &entry = special_attacks[aname];
	ptimeout = ...;
	if ( ptimeout >= 0 ) {
		entry.timeout = ptimeout;
	} else {
		entry.timeout = type->special_attacks.at(aname).cooldown;
		entry.enabled = false;
	}
```
This means no need for an explicit constructor.